### PR TITLE
Add the ability to disable automatically following redirects

### PR DESCRIPTION
**Pull request by annelo-msft**
Addresses https://github.com/Azure/azure-sdk-for-net/issues/32722

### CODE REVIEW

File: sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs [hunk 1]
```diff
@@ -8,17 +8,31 @@
 
 namespace Azure.Core.Pipeline
 {
-    internal sealed class RedirectPolicy : HttpPipelinePolicy
+    /// <summary>
+    /// A pipeline policy that detects a redirect response code and resends the request to the
+    /// location specified by the response.
+    /// </summary>
+    public sealed class RedirectPolicy : HttpPipelinePolicy
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs:15]: If we decide this is the approach we want to take in the Core API, I will move this file out of the `Internal` folder.  I wanted to leave it here for a clean diff in the initial PR.

File: sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs [hunk 2]
```diff
@@ -36,7 +50,9 @@ internal async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPi
             Request request = message.Request;
             Response response = message.Response;
 
-            while ((redirectUri = GetUriForRedirect(request, message.Response)) != null)
+            bool allowAutoRedirect = GetAllowAutoRedirectValue(message);
+
+            while (allowAutoRedirect && (redirectUri = GetUriForRedirect(request, message.Response)) != null)
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs:55]:
```suggestion
            if (!allowAutoRedirect) return;
            while ((redirectUri = GetUriForRedirect(request, message.Response)) != null)
```

File: sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs [hunk 3]
```diff
@@ -8,17 +8,31 @@
...
         {
             _maxAutomaticRedirections = 50;
         }
 
+        /// <summary>
+        /// Sets a value that indicates whether redirects will be automatically followed for this message.
+        /// </summary>
+        /// <param name="message"></param>
+        /// <param name="allowAutoRedirect"></param>
+        public static void AllowAutoRedirect(HttpMessage message, bool allowAutoRedirect)
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs:31]: Should this be called something like `SetAllowAutoRedirect` or `SetAutoRedirect`?

Comment (annelo-msft) [reply sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs:31]: I considered that ... I was worried that the former was wordy, and the latter didn't mirror the .NET API naming, but you're right, it should have a Set prefix.  I'll update it, and thanks!

File: sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs [hunk 4]
```diff
@@ -164,11 +180,20 @@ private static bool RequestRequiresForceGet(int statusCode, RequestMethod reques
         internal static bool IsSecureWebSocketScheme(string scheme) =>
             string.Equals(scheme, "wss", StringComparison.OrdinalIgnoreCase);
 
+        /// <summary>
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs:183]: nit: either fill out or inheritdoc

**Review by JoshLove-msft — APPROVED** [review]

File: sdk/containerregistry/Azure.Containers.ContainerRegistry/api/Azure.Containers.ContainerRegistry.netstandard2.0.cs [hunk 5]
```diff
@@ -138,6 +138,7 @@ public partial class ContainerRegistryClient
         public ContainerRegistryClient(System.Uri endpoint, Azure.Core.TokenCredential credential) { }
         public ContainerRegistryClient(System.Uri endpoint, Azure.Core.TokenCredential credential, Azure.Containers.ContainerRegistry.ContainerRegistryClientOptions options) { }
         public virtual System.Uri Endpoint { get { throw null; } }
+        public virtual Azure.Core.Pipeline.HttpPipeline Pipeline { get { throw null; } }
```
Comment (christothes) [inline sdk/containerregistry/Azure.Containers.ContainerRegistry/api/Azure.Containers.ContainerRegistry.netstandard2.0.cs:141]: Could you provide some context on why this needs to be public? If this behavior for a specific client call is intended to be exposed to customers, I'd expect it to be exposed via some client options or parameter on the specific methods that require it. If for the client itself, I'd expect it to be internal.

Comment (annelo-msft) [reply sdk/containerregistry/Azure.Containers.ContainerRegistry/api/Azure.Containers.ContainerRegistry.netstandard2.0.cs:141]: Sure, there are a couple of reasons.  

The first is that our team is moving toward a strategy of giving lower-level access to library users by exposing the Pipeline as a public member on the client.  This is generated automatically in DPG clients, so this is following that precedent, just implementing it manually in this case.

The second is that this scenario is sufficiently uncommon, and not fully documented by the service team, that we don't want to expose it in a supported API on a GA client, but we do want to make it available to callers who understand the details and are willing to use an advanced API to accomplish it.

File: sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs (hunk 1 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core/src/Pipeline/Internal/RedirectPolicy.cs:15]: This is completed.

File: sdk/containerregistry/Azure.Containers.ContainerRegistry/api/Azure.Containers.ContainerRegistry.netstandard2.0.cs (hunk 5 above)
Comment (annelo-msft) [reply sdk/containerregistry/Azure.Containers.ContainerRegistry/api/Azure.Containers.ContainerRegistry.netstandard2.0.cs:141]: Discussed offline.

**Pull request merged**